### PR TITLE
Kokkos Kernels: adding missing TPLs and pre-conditions

### PR DESCRIPTION
**Pull request by lucbv**
Adding variants and dependencies for rocBLAS and rocSPARSE. Also adding a "when=" close to the TPL variants that prevents enabling the TPLs in versions of the library when it was not yet available.

### CODE REVIEW

**Review by cedricchevalier19 — COMMENTED** [review]
Thanks for the information.
It's definitely a nice improvement.

**Comment by cedricchevalier19** [comment]
I let @brian-kelley approving as with Kokkos we are not consumers of Kokkos-Kernels.

**Review by brian-kelley — APPROVED** [review]

**Comment by lucbv** [comment]
Thanks @cedricchevalier19 and @brian-kelley let us wait for a Spack maintainer to have a look so we can merge this.
This will set us up a bit better to generate our modules on Frontier and Livermore's machines : )

**Comment by lucbv** [comment]
@alecbcs @tldahlgren could you have quick look at this please, it is ready to be merged

File: var/spack/repos/builtin/packages/kokkos-kernels/package.py [hunk 1]
```diff
@@ -136,19 +136,21 @@ class KokkosKernels(CMakePackage, CudaPackage):
 
     tpls = {
         # variant name   #deflt   #spack name   #root var name #docstring
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/kokkos-kernels/package.py]: Do you want to add the condition to this line so it matches the change?

File: var/spack/repos/builtin/packages/kokkos-kernels/package.py [hunk 2]
```diff
@@ -136,19 +136,21 @@ class KokkosKernels(CMakePackage, CudaPackage):
...
+        "lapacke": (False, "clapack", "LAPACKE", "@3.1.00:", "Link to LAPACKE library"),
+        "rocblas": (False, "rocblas", "ROCBLAS", "@3.6.00:", "Link to AMD BLAS library"),
+        "rocsparse": (False, "rocsparse", "ROCSPARSE", "@3.6.00:", "Link to AMD sparse library"),
     }
 
     for tpl in tpls:
-        deflt_bool, spackname, rootname, descr = tpls[tpl]
-        variant(tpl, default=deflt_bool, description=descr)
+        deflt_bool, spackname, rootname, condition, descr = tpls[tpl]
+        variant(tpl, default=deflt_bool, when="%s" % condition, description=descr)
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/kokkos-kernels/package.py]: FYI. You *could* use a python f-string here (see https://docs.python.org/3/tutorial/inputoutput.html#formatted-string-literals).  We're converting the core to use them as we make changes.

**Review by tldahlgren — DISMISSED** [review]
LGTM though I highly recommend adding the condition to the comments describing the contents of the tuple.

**Comment by tldahlgren** [comment]
Let me know if you want to merge as-is though I do recommend updating the comment first.

File: var/spack/repos/builtin/packages/kokkos-kernels/package.py (hunk 1 above)
Comment (lucbv) [reply var/spack/repos/builtin/packages/kokkos-kernels/package.py]: Yes, updated the comment to add the condition up there

File: var/spack/repos/builtin/packages/kokkos-kernels/package.py (hunk 2 above)
Comment (lucbv) [reply var/spack/repos/builtin/packages/kokkos-kernels/package.py]: Sure f-string is a bit simpler and more readable so switched to that. Assuming this all goes well I might create a later PR that goes through the CMake logic section of the package and switch it to f-string as well.

**Review by tldahlgren — APPROVED** [review]
Thanks.

**Pull request merged**